core/src/main/java/com/linecorp/armeria/server/RouteBuilder.java

```diff
@@ -267,6 +279,102 @@ public RouteBuilder produces(Iterable<MediaType> produceTypes) {
+     * <ul>
+     *     <li>{@code some-param=some-value} which the request must have a
+     *     {@code some-param=some-value} parameter</li>
+     *     <li>{@code some-param!=some-value} which the request must not have a
+     *     {@code some-param=some-value} parameter</li>
+     *     <li>{@code some-param} which the request must contain a {@code some-param} parameter</li>
+     *     <li>{@code !some-param} which the request must not contain a {@code some-param} parameter</li>
+     * </ul>
+     *
+     * <p>Note that these predicates can be evaluated only with the query string of the request URI.
+     * Also note that each predicate will be evaluated with the decoded value of HTTP parameters,
+     * so do not use percent-encoded value in the predicate.
+     *
+     * @see Params
+     */
+    public final RouteBuilder params(String... paramPredicates) {
+        return params(ImmutableList.copyOf(requireNonNull(paramPredicates, "paramPredicates")));
+    }
+
+    /**
+     * Sets the {@link Route} to accept a request if it matches all the specified predicates for
+     * HTTP parameters. The predicate can be one of the following forms:
+     * <ul>
+     *     <li>{@code some-param=some-value} which the request must have a
+     *     {@code some-param=some-value} parameter</li>
+     *     <li>{@code some-param!=some-value} which the request must not have a
+     *     {@code some-param=some-value} parameter</li>
+     *     <li>{@code some-param} which the request must contain a {@code some-param} parameter</li>
+     *     <li>{@code !some-param} which the request must not contain a {@code some-param} parameter</li>
+     * </ul>
+     *
+     * <p>Note that these predicates can be evaluated only with the query string of the request URI.
+     * Also note that each predicate will be evaluated with the decoded value of HTTP parameters,
+     * so do not use percent-encoded value in the predicate.
+     *
+     * @see Params
+     */
+    public RouteBuilder params(Iterable<String> paramPredicates) {
+        this.paramPredicates = ImmutableList.copyOf(requireNonNull(paramPredicates, "paramPredicates"));
+        return this;
+    }
+
+    /**
+     * Sets the {@link Route} to accept a request if it matches all the specified predicates for
+     * {@link HttpHeaders}. The predicate can be one of the following forms:
+     * <ul>
+     *     <li>{@code some-header=some-value} which the request must have a
+     *     {@code some-header: some-value} header</li>
+     *     <li>{@code some-header!=some-value} which the request must not have a
+     *     {@code some-header: some-value} header</li>
+     *     <li>{@code some-header} which the request must contain a {@code some-header} header</li>
+     *     <li>{@code !some-header} which the request must not contain a {@code some-header} header</li>
+     * </ul>
+     *
+     * @see Headers
+     */
+    public final RouteBuilder headers(String... headerPredicates) {
```

Comment:
`CharSequence...`? We represent headers as `AsciiString`s, so all header names must be converted to header names via `HttpHeaderNames.of()`.